Update openamp sample to use sysbuild

danieldegrasse opened the pull request:
Updates openamp sample to use sysbuild for all supported platforms.

For the LPC55s69 and LPC54114 SOCs, the method of flashing the second image was changed to use `CONFIG_BUILD_OUTPUT_ADJUST_LMA` instead of the custom solutions for combining images
both of these SOCs previously used.

the `ipm_mcux` sample was also updated to use sysbuild, as the legacy method of flashing both cores of the LPC55s69/54114 that this sample used was no longer supported with the updates made in this PR.

This PR has been tested on the LPC55s69 and LPC54114 EVKs, as those are the boards I have available that support the openamp sample. All other supported boards were verified to build

danieldegrasse committed:
boards: lpcxpresso55s69: use LMA adjustment for second core

Use LMA adjustment building dual core image with LPC55s69. The load
address adjustment allows the second core image to be flashed onto the chip
by the debugger, into the secure region of the LPC flash.

Additionally, remove the build dependency that was enforced by the SOC.
This requirement to build the secondary core image first can be enforced
by the project itself, instead of the core.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
samples: openamp: update openamp sample to use sysbuild

Update IPC openamp sample to use sysbuild for multi core images.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
samples: openamp: add additional required KConfigs for LPC55s69

LPC55s69 requires the second core to output a hex build to ensure the
image is flashed to the correct location, and uses the
CONFIG_SECOND_CORE_MCUX symbol in each core's build to indicate the build
targets a dual core config

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
soc: lpc54114: update multi core support for LMA address offset

Update lpc54114 support to use LMA address offset instead of linking
secondary core image into primary core memory. This will allow support
with sysbuild to be enabled. Additionally, use partitions to select where
the secondary core image will be located in flash

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
samples: openamp: add lpc54114 configuration settings

Add lpc54114 configuration settings to enable booting second core with
sysbuild, and LMA address adjustment

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
samples: openamp: add separate overlay files for remote cores

with sysbuild, DTS overlays for boards cannot automatically be passed to
the remote core. This means DTS overlays must be present for remote cores
as well.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
scripts: Fix header ifdef guard for gen_image_info.py

gen_image_info.py was generating invalid include guard. Properly strip
all special characters from filename to generate include guard #ifdef

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
samples: drivers: ipm: update to use sysbuild

Update ipm_mcux sample to use sysbuild for multicore images

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

tejlmand commented on samples/subsys/ipc/openamp/CMakeLists.txt:
```diff
@@ -5,7 +5,7 @@ cmake_minimum_required(VERSION 3.20.0)
 #
 
-set(REMOTE_ZEPHYR_DIR ${CMAKE_CURRENT_BINARY_DIR}/openamp_remote-prefix/src/openamp_remote-build/zephyr)
+set(REMOTE_ZEPHYR_DIR ${CMAKE_CURRENT_BINARY_DIR}/../openamp_remote/zephyr)
```
Somewhat unfortunate. It has to tightly coupled expectation on where the include dir of the second remote image is located.

It's knowledge like this we should be able to share in a generic way through sysbuild.
Will do some thinking on designing a better approach.

danieldegrasse replied:
Sounds good. Maybe we could expose this with a variable?

tejlmand replied:
I have wanted to fix this for some time: https://github.com/zephyrproject-rtos/zephyr/pull/49286
not solving this exact situation but laying down a foundation so we can easier handle a case like this.

danieldegrasse replied:
`zephyr_get` looks like it would solve this issue more generically. For the time being, will leave the current implementation as is, until that infrastructure is available.

tejlmand replied:
`zephyr_get()` PR #49674 has been merged.
Could you try to update this PR, and see if more functionality is needed.

I assume we need to split the `ExternalZephyrProject_Add()` into two stages.
Stage 1 which will handle the request for adding a Zephyr image, and then a later stage 2 that will invoke the CMake configuration of the image list.

This will allow us to know all images before running CMake, and thus allow us to share location of each image for CMake.

I don't believe this PR should be blocked by such improvement, as it is more important to transitionto sysbuild and gain knowledge of improvements to work on.

danieldegrasse replied:
I tested `zephyr_get()` with this PR, and I agree that it seems like we need a split like you described. Currently, the main sysbuild application must be aware of the build location of the remote application, but the primary sysbuild application is configured in the call to `ExternalZephyrProject_Add()`. This call occurs prior to the secondary application(s) being added via `sysbuild.cmake`

Maybe we could record all image names prior to running CMake configuration, and write these image names into a CMake cache file that all images can load variables from. We could then add this cache file into the list of sources used by `zephyr_get()`

tejlmand commented on samples/subsys/ipc/openamp/remote/boards/mps2_an521_remote.overlay:
```diff
@@ -0,0 +1,20 @@
+/*
```
nothing wrong on this line, but the commit message confuses me a bit.

> with sysbuild, kconfig overlays for boards cannot be passed to the remote
core. This means config file overlays must be present for remote cores as
well.

but here you are adding devicetree overlays ?
Secondly, sysbuild does support passing overlays to other builds (cores), like this:
`-D<image>_DTC_OVERLAY_FILE=<filename>`

but assume you are referring to automatically picking up board overlays for an image built as part of sysbuild.
Could you be a bit more precise on the message, and also feel free to comment if you think there is an extra use-case that sysbuild should support, or if you find this solution correct.

danieldegrasse replied:
I will update the commit message here to be correct, as the issue is that sysbuild does not allow automatically passing devicetree overlays between builds.

`-D<image>_DTC_OVERLAY_FILE=<filename>` does work correctly, but in this case the desired behavior is that the overlay CMake selects for the primary core is then automatically used for the secondary core. Since both projects are added as external projects, the `DTC_OVERLAY_FILE` variable for the main application cannot be accessed in `sysbuild.cmake`, and thus cannot be passed onto the build for the secondary core (except via commandline, as you mentioned)

tejlmand replied:
> the `DTC_OVERLAY_FILE` variable for the main application cannot be accessed in `sysbuild.cmake`, and thus cannot be passed onto the build for the secondary core

with `zephyr_get()` in place, this is actually possible.
https://github.com/zephyrproject-rtos/zephyr/pull/49674

we load the local sysbuild information for DTC_OVERLAY_FILE here:
https://github.com/zephyrproject-rtos/zephyr/blob/4a702ea21ef82b301464551807358f8490951e45/cmake/modules/configuration_files.cmake#L92

we could loosen up on this, so that an image may fetch the global overlay file, but we need to be really careful, cause in some cases images doesn't want to know about other images overlays.

So we should specify a clear pattern that is logical to users as to when such overlays are used be a single or multiple images.
Also we must consider that there can be multi images where two images wants the same overlay, but not the 3rd image.
(two cores where both want the overlay, and then a third app image on just one of the cores that doesn't want this overlay)

tejlmand commented on samples/subsys/ipc/openamp/sysbuild.cmake:
```diff
@@ -0,0 +1,29 @@
+#
+# Copyright 2022 NXP
+
+# Select correct remote target for a given board
```
I believe going through Kconfig here would be a cleaner approach.
It can couple the remote board name / dependency to the use of a second remote image.

It avoids several `if ... elseif ... elseif ... else` in CMake.
It allows users to extend the Kconfig downstream with extra remote board names without modifying Zephyr tree, for example like this:
```
# Downstream Kconfig.private:
config SB_BOARD_openamp_remote
	string
	default "my_remote_board" if $(BOARD) = "my_main_board"
```

And finally, it allow us to easily refactor this when we get SoC and boards Kconfig cleaned up in a completely generic part (that can be used in both Zephyr and sysbuild) and a Zephyr specific part.

Did you look at this prototype ?
https://github.com/tejlmand/zephyr/commit/efa77ac2e8507fea7bf4be4a67bd0b42af7899a7

especially:
https://github.com/tejlmand/zephyr/blob/efa77ac2e8507fea7bf4be4a67bd0b42af7899a7/samples/subsys/ipc/openamp/Kconfig.sysbuild
https://github.com/tejlmand/zephyr/blob/efa77ac2e8507fea7bf4be4a67bd0b42af7899a7/samples/subsys/ipc/openamp/sysbuild.cmake

danieldegrasse replied:
I've updated the PR to use Kconfig for the remote board name, based off the prototype you mentioned. Thanks for the heads up, I was unaware of `Kconfig.sysbuild`

tejlmand reviewed (changes requested):
Thanks a lot for taking up this :+1: 

It helps bring sysbuild forward.

tejlmand commented on samples/subsys/ipc/openamp/sysbuild.cmake:
```diff
@@ -0,0 +1,15 @@
+
+# Add a dependency so that the remote sample will be built and flashed first
+add_dependencies(openamp openamp_remote)
+# Place remote image first in the image list
```
could you provide a bit more information as to why exactly this is required ?

For example like it's done for MCUboot:
https://github.com/zephyrproject-rtos/zephyr/blob/4a702ea21ef82b301464551807358f8490951e45/share/sysbuild/bootloader/CMakeLists.txt#L11-L12

danieldegrasse replied:
I've updated the comment- the build dependency exists because the remote image needs to generate a build header, using `CONFIG_BUILD_OUTPUT_INFO_HEADER`. The primary image includes that header, so the remote image must be built first.

tejlmand commented on samples/drivers/ipm/ipm_mcux/sysbuild.cmake:
```diff
@@ -0,0 +1,23 @@
+#
+# Copyright 2022 NXP
+
+if("${BOARD}" STREQUAL "lpcxpresso54114_m4")
+  set(BOARD_REMOTE "lpcxpresso54114_m0")
+elseif("${BOARD}" STREQUAL "lpcxpresso55s69_cpu0")
+  set(BOARD_REMOTE "lpcxpresso55s69_cpu1")
+else()
+  message(FATAL_ERROR "${BOARD} was not supported for this sample")
+endif()
```
I believe this should also be updated to use the Kconfig approach.

That will allow downstream users to easily extend the Kconfig setting with downstream only remote board without patching Zephyr repo.

tejlmand reviewed (changes requested):
looks good.

Almost ready to go in.

carlescufi commented:
@danieldegrasse could you address the review comments please?

danieldegrasse commented:
> could you address the review comments please?

Comments have been addressed- wanted to hold off until #49552, because without sysbuild support in twister this PR would fail CI.

danieldegrasse commented:
This PR is now dependent on #51045 (or another PR with equivalent functionality), because twister needs to know how to run QEMU targets like `mps2_an521` to properly test this sample with sysbuild enabled

danieldegrasse committed:
boards: mps2_an521: add remote Zephyr dir for openamp

Add remote Zephyr directory for openamp sample, so that the sample
will correctly set emulator flags when compiled with sysbuild

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse commented:
@tejlmand Could you take another look at this PR, now that #51045 is merged and CI can pass?

carlocaione commented on soc/arm/nxp_lpc/lpc54xxx/soc.c:
```diff
@@ -27,6 +27,17 @@
+ * n is the segment number, as defined in zephyr_image_info.h
+ */
+#define MEMCPY_SEGMENT(n, _)							\
+	memcpy((uint32_t *)((SEGMENT_LMA_ADDRESS_ ## n) - ADJUSTED_LMA),	\
+		(uint32_t *)(SEGMENT_LMA_ADDRESS_ ## n),			\
+		(SEGMENT_SIZE_ ## n))
```
nit: you probably want to use `z_early_memcpy`

carlocaione reviewed (approved):
sysbuild is (currently) not my cup of tea but all the rest seems ok.

tejlmand reviewed (approved):
lgtm.

Thanks a lot for this update.
